### PR TITLE
elaboration performance improvement

### PR DESCRIPTION
**Pull request by jsn1993**
Ufortunately for HUGE designs I cannot really do much because there are 30k+ update blocks and the elaboration code involves a TON of for/if/isinstance statements, rending itself JIT-unfriendly

### CODE REVIEW

**Comment by jsn1993** [comment]
how do we actually fix those travis error due to hypothesis/attr?

File: pymtl3/dsl/Component.py [hunk 1]
```diff
@@ -73,7 +73,7 @@ def _collect_objects_local( s, filt ):
     stack = []
     for (name, obj) in s.__dict__.items():
       if   isinstance( name, str ):
-        if not name.startswith("_"): # filter private variables
+        if name[0] != '_': # filter private variables
```
Comment (ptpan) [inline pymtl3/dsl/Component.py]: can name be an empty string?

Comment (jsn1993) [reply pymtl3/dsl/Component.py]: i don't think so

File: pymtl3/dsl/Component.py [hunk 2]
```diff
@@ -490,15 +506,17 @@ def unlock_simulation( s ):
     except:
       raise AttributeError("Cannot unlock an unlocked/never locked model.")
 
+    swapped_values  = defaultdict(list)
```
Comment (ptpan) [inline pymtl3/dsl/Component.py]: Is this mostly a performance optimization that tries to remove unnecessary attribute lookups?

File: requirements.txt [hunk 3]
```diff
@@ -1,9 +1,7 @@
 pytest >= 4.4
 hypothesis >= 4.18.3
-pytest-xdist
 cffi
 greenlet
-pyparsing
 graphviz
```
Comment (ptpan) [inline requirements.txt]: Cool. Let's definitely do more clean up like this to make pip install much faster

**Comment by cbatten** [comment]
definitely! making pip install faster is a great goal

File: pymtl3/dsl/Component.py (hunk 2 above)
Comment (jsn1993) [reply pymtl3/dsl/Component.py]: Yes ... I try to only do it when it doesn't hurt readability ...

File: requirements.txt (hunk 3 above)
Comment (jsn1993) [reply requirements.txt]: I wonder if we can remove graphviz

Comment (ptpan) [reply requirements.txt]: IMO graphviz really should be a developer tool that facilitates debugging. We probably should have different "dependency groups" which expose different set of tools to different audience. For example the "product" group should have core dependencies only, whereas "dev" group might have extra debugging facilitates like graphviz. Python formatters and code coverage tools should probably also go to the "dev" group.

Comment (jsn1993) [reply requirements.txt]: hmm maybe we can move "CI dependencies" to .travis.yml?

**Pull request merged**